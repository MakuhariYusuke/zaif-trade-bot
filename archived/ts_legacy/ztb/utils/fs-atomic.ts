import fs from 'fs';
import path from 'path';

export function writeFileAtomic(filePath: string, data: string | Buffer): void {
  const dir = path.dirname(filePath);
  if (!fs.existsSync(dir)) fs.mkdirSync(dir, { recursive: true });
  const tmp = path.join(
    dir,
    `.${path.basename(filePath)}.tmp-${process.pid}-${Date.now()}-${Math.random()
      .toString(36)
      .slice(2)}`
  );
  const fd = fs.openSync(tmp, 'w');
  try {
    if (typeof data === 'string') {
      fs.writeFileSync(fd, data, 'utf8');
    } else {
      fs.writeFileSync(fd, data);
    }
    fs.fsyncSync(fd);
  } finally {
    try { fs.closeSync(fd); } catch { /* ignore */ }
  }
  fs.renameSync(tmp, filePath);
}
