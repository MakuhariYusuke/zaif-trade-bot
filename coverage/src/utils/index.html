
<!doctype html>
<html lang="en">

<head>
    <title>Code coverage report for src/utils</title>
    <meta charset="utf-8" />
    <link rel="stylesheet" href="../../prettify.css" />
    <link rel="stylesheet" href="../../base.css" />
    <link rel="shortcut icon" type="image/x-icon" href="../../favicon.png" />
    <meta name="viewport" content="width=device-width, initial-scale=1" />
    <style type='text/css'>
        .coverage-summary .sorter {
            background-image: url(../../sort-arrow-sprite.png);
        }
    </style>
</head>
    
<body>
<div class='wrapper'>
    <div class='pad1'>
        <h1><a href="../../index.html">All files</a> src/utils</h1>
        <div class='clearfix'>
            
            <div class='fl pad1y space-right2'>
                <span class="strong">31.23% </span>
                <span class="quiet">Statements</span>
                <span class='fraction'>253/810</span>
            </div>
        
            
            <div class='fl pad1y space-right2'>
                <span class="strong">15% </span>
                <span class="quiet">Branches</span>
                <span class='fraction'>6/40</span>
            </div>
        
            
            <div class='fl pad1y space-right2'>
                <span class="strong">21.42% </span>
                <span class="quiet">Functions</span>
                <span class='fraction'>6/28</span>
            </div>
        
            
            <div class='fl pad1y space-right2'>
                <span class="strong">31.23% </span>
                <span class="quiet">Lines</span>
                <span class='fraction'>253/810</span>
            </div>
        
            
        </div>
        <p class="quiet">
            Press <em>n</em> or <em>j</em> to go to the next uncovered block, <em>b</em>, <em>p</em> or <em>k</em> for the previous block.
        </p>
        <template id="filterTemplate">
            <div class="quiet">
                Filter:
                <input type="search" id="fileSearch">
            </div>
        </template>
    </div>
    <div class='status-line low'></div>
    <div class="pad1">
<table class="coverage-summary">
<thead>
<tr>
   <th data-col="file" data-fmt="html" data-html="true" class="file">File</th>
   <th data-col="pic" data-type="number" data-fmt="html" data-html="true" class="pic"></th>
   <th data-col="statements" data-type="number" data-fmt="pct" class="pct">Statements</th>
   <th data-col="statements_raw" data-type="number" data-fmt="html" class="abs"></th>
   <th data-col="branches" data-type="number" data-fmt="pct" class="pct">Branches</th>
   <th data-col="branches_raw" data-type="number" data-fmt="html" class="abs"></th>
   <th data-col="functions" data-type="number" data-fmt="pct" class="pct">Functions</th>
   <th data-col="functions_raw" data-type="number" data-fmt="html" class="abs"></th>
   <th data-col="lines" data-type="number" data-fmt="pct" class="pct">Lines</th>
   <th data-col="lines_raw" data-type="number" data-fmt="html" class="abs"></th>
</tr>
</thead>
<tbody><tr>
	<td class="file low" data-value="config.ts"><a href="config.ts.html">config.ts</a></td>
	<td data-value="0" class="pic low">
	<div class="chart"><div class="cover-fill" style="width: 0%"></div><div class="cover-empty" style="width: 100%"></div></div>
	</td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="59" class="abs low">0/59</td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="1" class="abs low">0/1</td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="1" class="abs low">0/1</td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="59" class="abs low">0/59</td>
	</tr>

<tr>
	<td class="file medium" data-value="daily-stats.ts"><a href="daily-stats.ts.html">daily-stats.ts</a></td>
	<td data-value="57.39" class="pic medium">
	<div class="chart"><div class="cover-fill" style="width: 57%"></div><div class="cover-empty" style="width: 43%"></div></div>
	</td>
	<td data-value="57.39" class="pct medium">57.39%</td>
	<td data-value="230" class="abs medium">132/230</td>
	<td data-value="25" class="pct low">25%</td>
	<td data-value="16" class="abs low">4/16</td>
	<td data-value="33.33" class="pct low">33.33%</td>
	<td data-value="12" class="abs low">4/12</td>
	<td data-value="57.39" class="pct medium">57.39%</td>
	<td data-value="230" class="abs medium">132/230</td>
	</tr>

<tr>
	<td class="file high" data-value="features-logger.ts"><a href="features-logger.ts.html">features-logger.ts</a></td>
	<td data-value="100" class="pic high">
	<div class="chart"><div class="cover-fill cover-full" style="width: 100%"></div><div class="cover-empty" style="width: 0%"></div></div>
	</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="75" class="abs high">75/75</td>
	<td data-value="5.88" class="pct low">5.88%</td>
	<td data-value="17" class="abs low">1/17</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="1" class="abs high">1/1</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="75" class="abs high">75/75</td>
	</tr>

<tr>
	<td class="file low" data-value="logger.ts"><a href="logger.ts.html">logger.ts</a></td>
	<td data-value="0" class="pic low">
	<div class="chart"><div class="cover-fill" style="width: 0%"></div><div class="cover-empty" style="width: 100%"></div></div>
	</td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="117" class="abs low">0/117</td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="1" class="abs low">0/1</td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="1" class="abs low">0/1</td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="117" class="abs low">0/117</td>
	</tr>

<tr>
	<td class="file low" data-value="price-cache.ts"><a href="price-cache.ts.html">price-cache.ts</a></td>
	<td data-value="0" class="pic low">
	<div class="chart"><div class="cover-fill" style="width: 0%"></div><div class="cover-empty" style="width: 100%"></div></div>
	</td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="53" class="abs low">0/53</td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="1" class="abs low">0/1</td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="1" class="abs low">0/1</td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="53" class="abs low">0/53</td>
	</tr>

<tr>
	<td class="file low" data-value="result.ts"><a href="result.ts.html">result.ts</a></td>
	<td data-value="0" class="pic low">
	<div class="chart"><div class="cover-fill" style="width: 0%"></div><div class="cover-empty" style="width: 100%"></div></div>
	</td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="6" class="abs low">0/6</td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="1" class="abs low">0/1</td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="1" class="abs low">0/1</td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="6" class="abs low">0/6</td>
	</tr>

<tr>
	<td class="file low" data-value="signer.ts"><a href="signer.ts.html">signer.ts</a></td>
	<td data-value="0" class="pic low">
	<div class="chart"><div class="cover-fill" style="width: 0%"></div><div class="cover-empty" style="width: 100%"></div></div>
	</td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="37" class="abs low">0/37</td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="1" class="abs low">0/1</td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="1" class="abs low">0/1</td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="37" class="abs low">0/37</td>
	</tr>

<tr>
	<td class="file low" data-value="toolkit.ts"><a href="toolkit.ts.html">toolkit.ts</a></td>
	<td data-value="32.16" class="pic low">
	<div class="chart"><div class="cover-fill" style="width: 32%"></div><div class="cover-empty" style="width: 68%"></div></div>
	</td>
	<td data-value="32.16" class="pct low">32.16%</td>
	<td data-value="143" class="abs low">46/143</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="1" class="abs high">1/1</td>
	<td data-value="11.11" class="pct low">11.11%</td>
	<td data-value="9" class="abs low">1/9</td>
	<td data-value="32.16" class="pct low">32.16%</td>
	<td data-value="143" class="abs low">46/143</td>
	</tr>

<tr>
	<td class="file low" data-value="trade-logger.ts"><a href="trade-logger.ts.html">trade-logger.ts</a></td>
	<td data-value="0" class="pic low">
	<div class="chart"><div class="cover-fill" style="width: 0%"></div><div class="cover-empty" style="width: 100%"></div></div>
	</td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="90" class="abs low">0/90</td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="1" class="abs low">0/1</td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="1" class="abs low">0/1</td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="90" class="abs low">0/90</td>
	</tr>

</tbody>
</table>
</div>
                <div class='push'></div><!-- for sticky footer -->
            </div><!-- /wrapper -->
            <div class='footer quiet pad2 space-top1 center small'>
                Code coverage generated by
                <a href="https://istanbul.js.org/" target="_blank" rel="noopener noreferrer">istanbul</a>
                at 2025-09-13T15:47:59.977Z
            </div>
        <script src="../../prettify.js"></script>
        <script>
            window.onload = function () {
                prettyPrint();
            };
        </script>
        <script src="../../sorter.js"></script>
        <script src="../../block-navigation.js"></script>
    </body>
</html>
    