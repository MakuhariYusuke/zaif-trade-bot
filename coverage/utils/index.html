
<!doctype html>
<html lang="en">

<head>
    <title>Code coverage report for utils</title>
    <meta charset="utf-8" />
    <link rel="stylesheet" href="../prettify.css" />
    <link rel="stylesheet" href="../base.css" />
    <link rel="shortcut icon" type="image/x-icon" href="../favicon.png" />
    <meta name="viewport" content="width=device-width, initial-scale=1" />
    <style type='text/css'>
        .coverage-summary .sorter {
            background-image: url(../sort-arrow-sprite.png);
        }
    </style>
</head>
    
<body>
<div class='wrapper'>
    <div class='pad1'>
        <h1><a href="../index.html">All files</a> utils</h1>
        <div class='clearfix'>
            
            <div class='fl pad1y space-right2'>
                <span class="strong">82.58% </span>
                <span class="quiet">Statements</span>
                <span class='fraction'>1152/1395</span>
            </div>
        
            
            <div class='fl pad1y space-right2'>
                <span class="strong">69.94% </span>
                <span class="quiet">Branches</span>
                <span class='fraction'>263/376</span>
            </div>
        
            
            <div class='fl pad1y space-right2'>
                <span class="strong">73.46% </span>
                <span class="quiet">Functions</span>
                <span class='fraction'>72/98</span>
            </div>
        
            
            <div class='fl pad1y space-right2'>
                <span class="strong">82.58% </span>
                <span class="quiet">Lines</span>
                <span class='fraction'>1152/1395</span>
            </div>
        
            
        </div>
        <p class="quiet">
            Press <em>n</em> or <em>j</em> to go to the next uncovered block, <em>b</em>, <em>p</em> or <em>k</em> for the previous block.
        </p>
        <template id="filterTemplate">
            <div class="quiet">
                Filter:
                <input type="search" id="fileSearch">
            </div>
        </template>
    </div>
    <div class='status-line high'></div>
    <div class="pad1">
<table class="coverage-summary">
<thead>
<tr>
   <th data-col="file" data-fmt="html" data-html="true" class="file">File</th>
   <th data-col="pic" data-type="number" data-fmt="html" data-html="true" class="pic"></th>
   <th data-col="statements" data-type="number" data-fmt="pct" class="pct">Statements</th>
   <th data-col="statements_raw" data-type="number" data-fmt="html" class="abs"></th>
   <th data-col="branches" data-type="number" data-fmt="pct" class="pct">Branches</th>
   <th data-col="branches_raw" data-type="number" data-fmt="html" class="abs"></th>
   <th data-col="functions" data-type="number" data-fmt="pct" class="pct">Functions</th>
   <th data-col="functions_raw" data-type="number" data-fmt="html" class="abs"></th>
   <th data-col="lines" data-type="number" data-fmt="pct" class="pct">Lines</th>
   <th data-col="lines_raw" data-type="number" data-fmt="html" class="abs"></th>
</tr>
</thead>
<tbody><tr>
	<td class="file medium" data-value="config.ts"><a href="config.ts.html">config.ts</a></td>
	<td data-value="55.07" class="pic medium">
	<div class="chart"><div class="cover-fill" style="width: 55%"></div><div class="cover-empty" style="width: 45%"></div></div>
	</td>
	<td data-value="55.07" class="pct medium">55.07%</td>
	<td data-value="69" class="abs medium">38/69</td>
	<td data-value="66.66" class="pct medium">66.66%</td>
	<td data-value="6" class="abs medium">4/6</td>
	<td data-value="25" class="pct low">25%</td>
	<td data-value="8" class="abs low">2/8</td>
	<td data-value="55.07" class="pct medium">55.07%</td>
	<td data-value="69" class="abs medium">38/69</td>
	</tr>

<tr>
	<td class="file high" data-value="daily-stats.ts"><a href="daily-stats.ts.html">daily-stats.ts</a></td>
	<td data-value="86.78" class="pic high">
	<div class="chart"><div class="cover-fill" style="width: 86%"></div><div class="cover-empty" style="width: 14%"></div></div>
	</td>
	<td data-value="86.78" class="pct high">86.78%</td>
	<td data-value="386" class="abs high">335/386</td>
	<td data-value="70.27" class="pct medium">70.27%</td>
	<td data-value="111" class="abs medium">78/111</td>
	<td data-value="63.63" class="pct medium">63.63%</td>
	<td data-value="22" class="abs medium">14/22</td>
	<td data-value="86.78" class="pct high">86.78%</td>
	<td data-value="386" class="abs high">335/386</td>
	</tr>

<tr>
	<td class="file high" data-value="features-logger.ts"><a href="features-logger.ts.html">features-logger.ts</a></td>
	<td data-value="88.78" class="pic high">
	<div class="chart"><div class="cover-fill" style="width: 88%"></div><div class="cover-empty" style="width: 12%"></div></div>
	</td>
	<td data-value="88.78" class="pct high">88.78%</td>
	<td data-value="205" class="abs high">182/205</td>
	<td data-value="77.92" class="pct medium">77.92%</td>
	<td data-value="77" class="abs medium">60/77</td>
	<td data-value="88.88" class="pct high">88.88%</td>
	<td data-value="9" class="abs high">8/9</td>
	<td data-value="88.78" class="pct high">88.78%</td>
	<td data-value="205" class="abs high">182/205</td>
	</tr>

<tr>
	<td class="file high" data-value="logger.ts"><a href="logger.ts.html">logger.ts</a></td>
	<td data-value="94.95" class="pic high">
	<div class="chart"><div class="cover-fill" style="width: 94%"></div><div class="cover-empty" style="width: 6%"></div></div>
	</td>
	<td data-value="94.95" class="pct high">94.95%</td>
	<td data-value="119" class="abs high">113/119</td>
	<td data-value="71.79" class="pct medium">71.79%</td>
	<td data-value="39" class="abs medium">28/39</td>
	<td data-value="69.23" class="pct medium">69.23%</td>
	<td data-value="13" class="abs medium">9/13</td>
	<td data-value="94.95" class="pct high">94.95%</td>
	<td data-value="119" class="abs high">113/119</td>
	</tr>

<tr>
	<td class="file high" data-value="price-cache.ts"><a href="price-cache.ts.html">price-cache.ts</a></td>
	<td data-value="93.22" class="pic high">
	<div class="chart"><div class="cover-fill" style="width: 93%"></div><div class="cover-empty" style="width: 7%"></div></div>
	</td>
	<td data-value="93.22" class="pct high">93.22%</td>
	<td data-value="118" class="abs high">110/118</td>
	<td data-value="83.33" class="pct high">83.33%</td>
	<td data-value="36" class="abs high">30/36</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="7" class="abs high">7/7</td>
	<td data-value="93.22" class="pct high">93.22%</td>
	<td data-value="118" class="abs high">110/118</td>
	</tr>

<tr>
	<td class="file high" data-value="result.ts"><a href="result.ts.html">result.ts</a></td>
	<td data-value="100" class="pic high">
	<div class="chart"><div class="cover-fill cover-full" style="width: 100%"></div><div class="cover-empty" style="width: 0%"></div></div>
	</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="6" class="abs high">6/6</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="2" class="abs high">2/2</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="2" class="abs high">2/2</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="6" class="abs high">6/6</td>
	</tr>

<tr>
	<td class="file high" data-value="signer.ts"><a href="signer.ts.html">signer.ts</a></td>
	<td data-value="100" class="pic high">
	<div class="chart"><div class="cover-fill cover-full" style="width: 100%"></div><div class="cover-empty" style="width: 0%"></div></div>
	</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="37" class="abs high">37/37</td>
	<td data-value="83.33" class="pct high">83.33%</td>
	<td data-value="6" class="abs high">5/6</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="5" class="abs high">5/5</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="37" class="abs high">37/37</td>
	</tr>

<tr>
	<td class="file medium" data-value="test-reset.ts"><a href="test-reset.ts.html">test-reset.ts</a></td>
	<td data-value="79.68" class="pic medium">
	<div class="chart"><div class="cover-fill" style="width: 79%"></div><div class="cover-empty" style="width: 21%"></div></div>
	</td>
	<td data-value="79.68" class="pct medium">79.68%</td>
	<td data-value="64" class="abs medium">51/64</td>
	<td data-value="41.66" class="pct low">41.66%</td>
	<td data-value="12" class="abs low">5/12</td>
	<td data-value="66.66" class="pct medium">66.66%</td>
	<td data-value="3" class="abs medium">2/3</td>
	<td data-value="79.68" class="pct medium">79.68%</td>
	<td data-value="64" class="abs medium">51/64</td>
	</tr>

<tr>
	<td class="file medium" data-value="toolkit.ts"><a href="toolkit.ts.html">toolkit.ts</a></td>
	<td data-value="71.09" class="pic medium">
	<div class="chart"><div class="cover-fill" style="width: 71%"></div><div class="cover-empty" style="width: 29%"></div></div>
	</td>
	<td data-value="71.09" class="pct medium">71.09%</td>
	<td data-value="301" class="abs medium">214/301</td>
	<td data-value="56.57" class="pct medium">56.57%</td>
	<td data-value="76" class="abs medium">43/76</td>
	<td data-value="78.94" class="pct medium">78.94%</td>
	<td data-value="19" class="abs medium">15/19</td>
	<td data-value="71.09" class="pct medium">71.09%</td>
	<td data-value="301" class="abs medium">214/301</td>
	</tr>

<tr>
	<td class="file medium" data-value="trade-logger.ts"><a href="trade-logger.ts.html">trade-logger.ts</a></td>
	<td data-value="73.33" class="pic medium">
	<div class="chart"><div class="cover-fill" style="width: 73%"></div><div class="cover-empty" style="width: 27%"></div></div>
	</td>
	<td data-value="73.33" class="pct medium">73.33%</td>
	<td data-value="90" class="abs medium">66/90</td>
	<td data-value="72.72" class="pct medium">72.72%</td>
	<td data-value="11" class="abs medium">8/11</td>
	<td data-value="80" class="pct high">80%</td>
	<td data-value="10" class="abs high">8/10</td>
	<td data-value="73.33" class="pct medium">73.33%</td>
	<td data-value="90" class="abs medium">66/90</td>
	</tr>

</tbody>
</table>
</div>
                <div class='push'></div><!-- for sticky footer -->
            </div><!-- /wrapper -->
            <div class='footer quiet pad2 space-top1 center small'>
                Code coverage generated by
                <a href="https://istanbul.js.org/" target="_blank" rel="noopener noreferrer">istanbul</a>
                at 2025-09-14T13:15:01.839Z
            </div>
        <script src="../prettify.js"></script>
        <script>
            window.onload = function () {
                prettyPrint();
            };
        </script>
        <script src="../sorter.js"></script>
        <script src="../block-navigation.js"></script>
    </body>
</html>
    