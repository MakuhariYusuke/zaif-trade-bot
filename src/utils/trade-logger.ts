import fs from "fs";
import path from "path";

const LOG_DIR = process.env.LOG_DIR || path.resolve(process.cwd(), "logs");

function dailyLogPath(dateStr?: string) {
    const d = dateStr || new Date().toISOString().slice(0, 10);
    return path.join(LOG_DIR, `trades-${d}.log`);
}

let currentDate = new Date().toISOString().slice(0, 10);
let currentLogPath = dailyLogPath(currentDate);

function rotateIfNeeded() {
    const today = new Date().toISOString().slice(0, 10);
    if (today !== currentDate) {
        currentDate = today;
        currentLogPath = dailyLogPath(today);
    }
}

function ensureDir() {
    fs.mkdirSync(LOG_DIR, { recursive: true });
}

export type TradeLogType = "SIGNAL" | "ORDER" | "EXECUTION" | "ERROR" | "INFO";

export interface TradeLogEntry {
    ts: string;           // ISO timestamp
    type: TradeLogType;
    message: string;
    data?: any;
}

export function logTrade(entry: TradeLogEntry) {
    try {
        rotateIfNeeded();
        ensureDir();
        fs.appendFileSync(currentLogPath, JSON.stringify(entry) + "\n");
    } catch (err) {
        console.error("Failed to write trade log:", err);
    }
}

export function logSignal(message: string, data?: any) {
    logTrade({ ts: new Date().toISOString(), type: "SIGNAL", message, data });
}
export function logOrder(message: string, data?: any) {
    logTrade({ ts: new Date().toISOString(), type: "ORDER", message, data });
}
export function logExecution(message: string, data?: any) {
    logTrade({ ts: new Date().toISOString(), type: "EXECUTION", message, data });
}
export function logTradeError(message: string, data?: any) {
    logTrade({ ts: new Date().toISOString(), type: "ERROR", message, data });
}
export function logTradeInfo(message: string, data?: any) {
    logTrade({ ts: new Date().toISOString(), type: "INFO", message, data });
}

export interface DailyReport {
    date: string;
    trades: number;
    signals: number;
    pnlEstimate: number; // 簡易 (約定ログが無いため推定用)
}

/**
 * Generate a daily report for the given date.
 * @param {string} date - The date string in 'YYYY-MM-DD' format.
 * @returns {DailyReport} The daily report for the specified date.
 */
export function generateDailyReport(date: string): DailyReport {
    ensureDir();
    const file = dailyLogPath(date);
    if (!fs.existsSync(file)) return { date, trades: 0, signals: 0, pnlEstimate: 0 };
    const content = fs.readFileSync(file, "utf8").trim();
    if (!content || content === "") return { date, trades: 0, signals: 0, pnlEstimate: 0 };
    const lines = content.split(/\n+/).filter(line => line.trim() !== "");
    let trades = 0, signals = 0, pnl = 0;
    for (const line of lines) {
        try {
            const e = JSON.parse(line) as TradeLogEntry;
            if (e.type === "SIGNAL") signals++;
            if (e.type === "EXECUTION") trades++;
            if (e.type === "EXECUTION" && e.data && typeof e.data.pnl === "number") pnl += e.data.pnl;
        } catch { /* ignore */ }
    }
    return { date, trades, signals, pnlEstimate: pnl };
}