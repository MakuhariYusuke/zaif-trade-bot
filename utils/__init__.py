from .notify import DiscordNotifier

__all__ = ['DiscordNotifier']