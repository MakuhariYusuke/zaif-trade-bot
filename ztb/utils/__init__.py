from .logger import LoggerManager
from .stats import calculate_skew, calculate_kurtosis, nan_ratio, correlation
from .report_generator import ReportGenerator
from .ci_utils import collect_ci_metrics, notify_ci_results
from .checkpoint import CheckpointManager

__all__ = [
    'LoggerManager',
    'calculate_skew', 'calculate_kurtosis', 'nan_ratio', 'correlation',
    'ReportGenerator',
    'collect_ci_metrics', 'notify_ci_results',
    'CheckpointManager'
]