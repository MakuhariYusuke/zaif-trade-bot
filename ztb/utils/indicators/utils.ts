// Pure indicator functions (no side effects)
// Lightweight implementations with optional previous-state to enable rolling updates.

export type Num = number | null;

// ---- Utils ----

/**
 * Simple memoization helper for expensive computations
 */
function memoize<T extends (...args: any[]) => any>(fn: T, keyFn?: (...args: Parameters<T>) => string): T {
  const cache = new Map<string, ReturnType<T>>();
  return ((...args: Parameters<T>) => {
    const key = keyFn ? keyFn(...args) : JSON.stringify(args);
    if (cache.has(key)) return cache.get(key)!;
    const result = fn(...args);
    cache.set(key, result);
    return result;
  }) as T;
}

/**
 * A helper function for Wilder's smoothing (a specific type of EMA with alpha = 1/period).
 * @param prevValue The previous smoothed value.
 * @param currentValue The current raw value.
 * @param period The smoothing period.
 * @returns The new smoothed value.
 */
export function wilderSmooth(prevValue: number, currentValue: number, period: number): number {
  return (prevValue * (period - 1) + currentValue) / period;
}

/**
 * Calculates the rolling maximum value over a specified window period.
 * @param values - Array of numeric values.
 * @param period - The size of the rolling window.
 * @returns An array where each element is the maximum value in the window, or null if not enough data.
 */
function rollingMax(values: number[], period: number): Array<number | null> {
  const out: Array<number | null> = new Array(values.length).fill(null);
  if (period <= 0) return out;
  const deque: number[] = [];
  for (let i = 0; i < values.length; i++) {
    while (deque.length && deque[0] <= i - period) deque.shift();
    while (deque.length && values[deque[deque.length - 1]] <= values[i]) deque.pop();
    deque.push(i);
    if (i >= period - 1) out[i] = values[deque[0]];
  }
  return out;
}

function rollingMin(values: number[], period: number): Array<number | null> {
  const out: Array<number | null> = new Array(values.length).fill(null);
  if (period <= 0) return out;
  const deque: number[] = [];
  for (let i = 0; i < values.length; i++) {
    while (deque.length && deque[0] <= i - period) deque.shift();
    while (deque.length && values[deque[deque.length - 1]] >= values[i]) deque.pop();
    deque.push(i);
    if (i >= period - 1) out[i] = values[deque[0]];
  }
  return out;
}

// Memoized versions for performance
export const rollingMaxMemo = memoize(rollingMax, (values, period) => `${period}-${JSON.stringify(values)}`);
export const rollingMinMemo = memoize(rollingMin, (values, period) => `${period}-${JSON.stringify(values)}`);

// Export rollingArgMax and rollingArgMin
export const rollingArgMax = memoize(_rollingArgMax, (values: number[], period: number) => `${period}-${JSON.stringify(values)}`);
export const rollingArgMin = memoize(_rollingArgMin, (values: number[], period: number) => `${period}-${JSON.stringify(values)}`);

/**
 * Calculates the rolling sum over a specified window period.
 * @param values - Array of numeric values.
 * @param period - The size of the rolling window.
 * @returns An array where each element is the sum of values in the window, or null if not enough data.
 */
function rollingSum(values: number[], period: number): Array<number | null> {
  const n = values.length;
  const out: Array<number | null> = new Array(n).fill(null);
  let sum = 0;
  for (let i = 0; i < n; i++) {
    sum += values[i];
    if (i >= period) sum -= values[i - period];
    if (i >= period - 1) out[i] = sum;
  }
  return out;
}

/**
 * Rolling Arg Max
 * Calculates the index of the maximum value in a rolling window.
 * @param values - Array of numeric values.
 * @param period - The size of the rolling window.
 * @returns An array of indices corresponding to the maximum values in each window.
 */
function _rollingArgMax(values: number[], period: number): Array<number | null> {
  const n = values.length;
  const out: Array<number | null> = new Array(n).fill(null);
  if (period <= 0) {
    // For period <= 0, return indices directly
    for (let i = 0; i < n; i++) out[i] = i;
    return out;
  }
  const dq: number[] = [];
  for (let i = 0; i < n; i++) {
    while (dq.length && dq[0] <= i - period) dq.shift();
    while (dq.length && values[dq[dq.length - 1]] <= values[i]) dq.pop();
    dq.push(i);
    if (i >= period - 1) out[i] = dq[0];
  }
  return out;
}

/**
 * Rolling Arg Min
 * Calculates the index of the minimum value in a rolling window.
 * @param values - Array of numeric values.
 * @param period - The size of the rolling window.
 * @returns An array of indices corresponding to the minimum values in each window.
 */
function _rollingArgMin(values: number[], period: number): Array<number | null> {
  const n = values.length;
  const out: Array<number | null> = new Array(n).fill(null);
  if (period <= 0) {
    // For period <= 0, return indices directly
    for (let i = 0; i < n; i++) out[i] = i;
    return out;
  }
  const dq: number[] = [];
  for (let i = 0; i < n; i++) {
    while (dq.length && dq[0] <= i - period) dq.shift();
    while (dq.length && values[dq[dq.length - 1]] >= values[i]) dq.pop();
    dq.push(i);
    if (i >= period - 1) out[i] = dq[0];
  }
  return out;
}