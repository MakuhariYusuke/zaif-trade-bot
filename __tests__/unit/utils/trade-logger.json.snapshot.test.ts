import { describe, it, beforeEach, expect } from 'vitest';
import fs from 'fs';
import path from 'path';
import { logTradeError, logTradeInfo, logSignal, generateDailyReport } from '../../../src/utils/trade-logger';

describe('trade-logger JSON mode snapshot', () => {
  let tmpDir: string;
  beforeEach(()=>{
    tmpDir = fs.mkdtempSync(path.join(process.cwd(), 'tmp-trade-log-json-'));
    process.env.LOG_DIR = tmpDir;
    process.env.LOG_JSON = '1';
  });

  it('writes JSON lines (INFO/ERROR/SIGNAL) snapshot stable', async () => {
    logSignal('strategy_entry', { side: 'buy', strength: 0.8 });
    logTradeInfo('queued_order', { id: 'OID123', qty: 0.01 });
    logTradeError('failed_execution', { reason: 'timeout', retry: true });
    const today = new Date().toISOString().slice(0,10);
    const file = path.join(tmpDir, `trades-${today}.log`);
    for (let i=0;i<40 && !fs.existsSync(file);i++) await new Promise(r=>setTimeout(r,5));
    expect(fs.existsSync(file)).toBe(true);
    const raw = fs.readFileSync(file, 'utf8').trim().split(/\n+/).filter(Boolean);
    const norm = raw.map(l=>{ const o = JSON.parse(l); o.ts = 'T'; return o; });
    expect(norm).toMatchSnapshot();
    const rep = generateDailyReport(today);
    expect(rep.signals).toBe(1);
  });
});
