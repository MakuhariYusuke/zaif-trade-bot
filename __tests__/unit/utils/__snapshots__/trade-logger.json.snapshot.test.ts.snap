// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`trade-logger JSON mode snapshot > writes JSON lines (INFO/ERROR/SIGNAL) snapshot stable 1`] = `
[
  {
    "data": {
      "side": "buy",
      "strength": 0.8,
    },
    "message": "strategy_entry",
    "ts": "T",
    "type": "SIGNAL",
  },
  {
    "data": {
      "id": "OID123",
      "qty": 0.01,
    },
    "message": "queued_order",
    "ts": "T",
    "type": "INFO",
  },
  {
    "data": {
      "reason": "timeout",
      "retry": true,
    },
    "message": "failed_execution",
    "ts": "T",
    "type": "ERROR",
  },
]
`;
